libyul/optimiser/UnusedFunctionsCommon.h

```diff
@@ -0,0 +1,172 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+
+#include <liblangutil/SourceLocation.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <variant>
+
+namespace solidity::yul
+{
+
+template<typename T>
+std::vector<T> applyBooleanMask(std::vector<T> const& _vec, std::vector<bool> const& _mask)
+{
+	yulAssert(_vec.size() == _mask.size(), "");
+
+	std::vector<T> ret;
+
+	for (size_t i = 0; i < _mask.size(); ++i)
+		if (_mask[i])
+			ret.push_back(_vec[i]);
+
+	return ret;
+}
+
+/// Find functions whose arguments are not used in its body. Also, find functions whose body
+/// satisfies a heuristic about pruning.
+bool wasPruned(FunctionDefinition const& _f, Dialect const& _dialect)
+{
+	// We skip the function body if it
+	// 1. is empty, or
+	// 2. is a single statement that is an assignment statement whose value is a non-builtin
```

Comment:
Does it actually make a big difference whether it is a builtin function call or not?

Reply:
I'm not sure about the impact. If the inliner can handle them, then we don't have to compare with the dialect. However, the inliner won't be able to handle, e.g., `function f(a) -> y { y :=  mload(f(1)) }` because it is recursive.

We could replace this simply by `return _f.body.statements().size() <= 1`

Reply:
What about `_f.body.statements().size() <= 1 && CodeSize::codeSize(_f.body) <= 1`? The second part is the condition in the full inliner.